File: LineScope/Model/Session.swift
import AppKit
import ImageIO
import LineScopeCore
import SwiftUI
import UniformTypeIdentifiers

struct Profile {
    struct Point: Identifiable {
        let id: Int
        let x: Double
        let y: Double
    }

    var points: [Point]
    var values: [Float]
    var samples: LineSamples
}

struct SpectrumPoint: Identifiable {
    let id: Int
    let frequency: Double
    let magnitude: Double
}

/// All state for one document window: the variants, the shared line, and view settings.
@MainActor @Observable
final class Session {
    let fileName: String
    let sourceInfo: SourceInfo

    private(set) var variants: [ImageVariant]
    var line: LineSegment = .initial

    var colorModel: ColorModel = .rgba {
        didSet { if channelIndex >= colorModel.channels.count { channelIndex = 0 } }
    }
    var channelIndex = 0
    var window: WindowFunction = .hann
    var removeMean = true
    var logMagnitude = true
    var frequencyAxis: FrequencyAxis = .cyclesPerPixel
    var useDevicePixels = false

    /// Panes per zone, indexed by `Zone.rawValue`.
    var panes: [[PaneState]]
    /// The variant operations apply to and the inspector describes.
    var focusedVariantID: UUID
    var pendingRequest: OperationRequest?
    private(set) var runningOperations = 0
    var errorMessage: String?

    init(document: ImageDocument, fileName: String) {
        let original = ImageVariant(
            buffer: document.buffer, cgImage: document.displayImage,
            parentID: nil, operation: nil, lineage: []
        )
        self.fileName = fileName
        self.sourceInfo = document.info
        self.variants = [original]
        self.focusedVariantID = original.id
        self.panes = Zone.allCases.map { _ in [PaneState(variantID: original.id)] }
    }

    var original: ImageVariant { variants[0] }

    var focusedVariant: ImageVariant {
        variant(focusedVariantID) ?? original
    }

    func variant(_ id: UUID) -> ImageVariant? {
        variants.first { $0.id == id }
    }

    var channelName: String { colorModel.channels[channelIndex] }

    // MARK: Panes

    func select(_ variantID: UUID, zone: Zone, pane: Int) {
        panes[zone.rawValue][pane].variantID = variantID
        focusedVariantID = variantID
    }

    func split(_ zone: Zone) {
        let shown = Set(panes[zone.rawValue].map(\.variantID))
        let next = variants.first { !shown.contains($0.id) } ?? focusedVariant
        panes[zone.rawValue].append(PaneState(variantID: next.id))
    }

    func unsplit(_ zone: Zone) {
        guard panes[zone.rawValue].count > 1 else { return }
        panes[zone.rawValue].removeLast()
    }

    // MARK: Variants

    func closeVariant(_ id: UUID) {
        guard let index = variants.firstIndex(where: { $0.id == id }), index > 0 else { return }
        variants.remove(at: index)
        for z in panes.indices {
            for p in panes[z].indices where panes[z][p].variantID == id {
                panes[z][p].variantID = original.id
            }
        }
        if focusedVariantID == id { focusedVariantID = original.id }
    }

    /// Applies an operation to the focused variant in the background and adds the result as a new tab.
    func perform(_ operation: ImageOperation) {
        let source = focusedVariant
        let input = source.buffer
        runningOperations += 1
        Task {
            let output = await Task.detached(priority: .userInitiated) {
                operation.apply(to: input)
            }.value
            runningOperations -= 1
            guard let cg = output.makeCGImage() else {
                errorMessage = "Could not render the result of \(operation.displayName)."
                return
            }
            let variant = ImageVariant(
                buffer: output, cgImage: cg, parentID: source.id,
                operation: operation, lineage: source.lineage + [operation]
            )
            variants.append(variant)
            show(variant.id)
        }
    }

    /// Shows a new variant in every zone: in the last pane when split (so the left pane keeps the
    /// comparison baseline), otherwise in the only pane.
    private func show(_ id: UUID) {
        for z in panes.indices {
            panes[z][panes[z].count - 1].variantID = id
        }
        focusedVariantID = id
    }

    // MARK: Analysis

    func profile(for variant: ImageVariant) -> Profile {
        let samples = LineSampler.sample(variant.buffer, along: line)
        let values = samples.colors.map {
            ColorConversion.value(of: $0, model: colorModel, channel: channelIndex)
        }
        let points = values.indices.map { Profile.Point(id: $0, x: samples.positions[$0], y: Double(values[$0])) }
        return Profile(points: points, values: values, samples: samples)
    }

    func spectrum(for profile: Profile) -> SpectrumResult {
        SpectrumAnalyzer.compute(
            profile.values, spacing: max(profile.samples.spacing, 1e-9),
            window: window, removeMean: removeMean
        )
    }

    func spectrumPoints(_ s: SpectrumResult) -> [SpectrumPoint] {
        let freqs = frequencyAxis == .cyclesPerPixel ? s.cyclesPerPixel : s.cyclesPerLine
        return s.amplitudes.indices.map { i in
            let a = s.amplitudes[i]
            return SpectrumPoint(
                id: i, frequency: freqs[i],
                magnitude: logMagnitude ? max(20 * log10(max(a, 1e-12)), Self.dBFloor) : a
            )
        }
    }

    static let dBFloor = -100.0

    // MARK: Export

    func exportFocused() {
        let variant = focusedVariant
        let panel = NSSavePanel()
        panel.allowedContentTypes = [.png]
        let base = (fileName as NSString).deletingPathExtension
        let suffix = variant.isOriginal ? "original" : variant.name
            .replacingOccurrences(of: " → ", with: "_")
            .replacingOccurrences(of: "/", with: "-")
        panel.nameFieldStringValue = "\(base)-\(suffix).png"
        guard panel.runModal() == .OK, let url = panel.url else { return }
        guard
            let dest = CGImageDestinationCreateWithURL(url as CFURL, UTType.png.identifier as CFString, 1, nil)
        else {
            errorMessage = "Could not create \(url.lastPathComponent)."
            return
        }
        CGImageDestinationAddImage(dest, variant.cgImage, nil)
        if !CGImageDestinationFinalize(dest) {
            errorMessage = "Could not write \(url.lastPathComponent)."
        }
    }
}

extension FocusedValues {
    @Entry var session: Session?
}

/// The session of the most recently key document window, for the inspector window (which is itself
/// key while being used, so focused values would not reach it).
@MainActor @Observable
final class ActiveSession {
    static let shared = ActiveSession()
    var session: Session?
}

File: Windows/LineScope.App/Model/Session.cs
using System.IO;
using System.Windows;
using System.Windows.Media.Imaging;
using LineScope.Core;
using Microsoft.Win32;

namespace LineScope.App.Model;

public sealed record Profile(double[] X, float[] Values, LineSamples Samples);

/// <summary>What changed in a <see cref="Session"/>, so views redraw only what they show.</summary>
[Flags]
public enum SessionChange
{
    None = 0,
    Line = 1,
    Settings = 2,
    Variants = 4,
    Panes = 8,
    Focus = 16,
    Running = 32,
    All = Line | Settings | Variants | Panes | Focus | Running,
}

/// <summary>All state for one document window: the variants, the shared line, and view settings.</summary>
public sealed class Session
{
    public string FileName { get; }
    public SourceInfo SourceInfo { get; }

    private readonly List<ImageVariant> _variants;
    public IReadOnlyList<ImageVariant> Variants => _variants;

    /// <summary>Raised on the UI thread after any state change.</summary>
    public event Action<SessionChange>? Changed;
    public event Action<string>? Error;

    public Session(ImageDocument document, string fileName)
    {
        var original = new ImageVariant(document.Buffer, document.DisplayImage, null, null, []);
        FileName = fileName;
        SourceInfo = document.Info;
        _variants = [original];
        _focusedVariantId = original.Id;
        Panes = ZoneInfo.All.Select(_ => new List<PaneState> { new(original.Id) }).ToArray();
    }

    private void Raise(SessionChange change) => Changed?.Invoke(change);

    private LineSegment _line = LineSegment.Initial;
    public LineSegment Line
    {
        get => _line;
        set { if (_line != value) { _line = value; Raise(SessionChange.Line); } }
    }

    private ColorModel _colorModel = ColorModel.Rgba;
    public ColorModel ColorModel
    {
        get => _colorModel;
        set
        {
            if (_colorModel == value) return;
            _colorModel = value;
            if (_channelIndex >= value.Channels().Length) _channelIndex = 0;
            Raise(SessionChange.Settings);
        }
    }

    private int _channelIndex;
    public int ChannelIndex
    {
        get => _channelIndex;
        set { if (_channelIndex != value) { _channelIndex = value; Raise(SessionChange.Settings); } }
    }

    private WindowFunction _window = WindowFunction.Hann;
    public WindowFunction Window
    {
        get => _window;
        set { if (_window != value) { _window = value; Raise(SessionChange.Settings); } }
    }

    private bool _removeMean = true;
    public bool RemoveMean
    {
        get => _removeMean;
        set { if (_removeMean != value) { _removeMean = value; Raise(SessionChange.Settings); } }
    }

    private bool _logMagnitude = true;
    public bool LogMagnitude
    {
        get => _logMagnitude;
        set { if (_logMagnitude != value) { _logMagnitude = value; Raise(SessionChange.Settings); } }
    }

    private FrequencyAxis _frequencyAxis = FrequencyAxis.CyclesPerPixel;
    public FrequencyAxis FrequencyAxis
    {
        get => _frequencyAxis;
        set { if (_frequencyAxis != value) { _frequencyAxis = value; Raise(SessionChange.Settings); } }
    }

    private bool _useDevicePixels;
    public bool UseDevicePixels
    {
        get => _useDevicePixels;
        set { if (_useDevicePixels != value) { _useDevicePixels = value; Raise(SessionChange.Settings); } }
    }

    /// <summary>Panes per zone, indexed by <see cref="Zone"/>.</summary>
    public List<PaneState>[] Panes { get; }

    private Guid _focusedVariantId;
    /// <summary>The variant operations apply to and the inspector describes.</summary>
    public Guid FocusedVariantId
    {
        get => _focusedVariantId;
        set { if (_focusedVariantId != value) { _focusedVariantId = value; Raise(SessionChange.Focus); } }
    }

    public int RunningOperations { get; private set; }

    public ImageVariant Original => _variants[0];
    public ImageVariant FocusedVariant => Variant(FocusedVariantId) ?? Original;
    public ImageVariant? Variant(Guid id) => _variants.FirstOrDefault(v => v.Id == id);
    public string ChannelName => ColorModel.Channels()[ChannelIndex];

    // Panes

    public void Select(Guid variantId, Zone zone, int pane)
    {
        Panes[(int)zone][pane].VariantId = variantId;
        _focusedVariantId = variantId;
        Raise(SessionChange.Panes | SessionChange.Focus);
    }

    public void Split(Zone zone)
    {
        var shown = Panes[(int)zone].Select(p => p.VariantId).ToHashSet();
        var next = _variants.FirstOrDefault(v => !shown.Contains(v.Id)) ?? FocusedVariant;
        Panes[(int)zone].Add(new PaneState(next.Id));
        Raise(SessionChange.Panes);
    }

    public void Unsplit(Zone zone)
    {
        var panes = Panes[(int)zone];
        if (panes.Count <= 1) return;
        panes.RemoveAt(panes.Count - 1);
        Raise(SessionChange.Panes);
    }

    // Variants

    public void CloseVariant(Guid id)
    {
        int index = _variants.FindIndex(v => v.Id == id);
        if (index <= 0) return;
        _variants.RemoveAt(index);
        foreach (var zone in Panes)
            foreach (var pane in zone.Where(p => p.VariantId == id))
                pane.VariantId = Original.Id;
        if (_focusedVariantId == id) _focusedVariantId = Original.Id;
        Raise(SessionChange.Variants | SessionChange.Panes | SessionChange.Focus);
    }

    /// <summary>Applies an operation to the focused variant in the background and adds the result as a new tab.</summary>
    public async void Perform(ImageOperation operation)
    {
        var source = FocusedVariant;
        var input = source.Buffer;
        RunningOperations++;
        Raise(SessionChange.Running);
        try
        {
            var (output, bitmap) = await Task.Run(() =>
            {
                var result = operation.Apply(input);
                return (result, Bitmaps.FromBuffer(result));
            });
            var variant = new ImageVariant(output, bitmap, source.Id, operation, [.. source.Lineage, operation]);
            _variants.Add(variant);
            Show(variant.Id);
        }
        catch (Exception ex)
        {
            Error?.Invoke($"Could not compute {operation.DisplayName}: {ex.Message}");
        }
        finally
        {
            RunningOperations--;
            Raise(SessionChange.Running);
        }
    }

    /// <summary>
    /// Shows a new variant in every zone: in the last pane when split (so the left pane keeps the
    /// comparison baseline), otherwise in the only pane.
    /// </summary>
    private void Show(Guid id)
    {
        foreach (var zone in Panes) zone[^1].VariantId = id;
        _focusedVariantId = id;
        Raise(SessionChange.Variants | SessionChange.Panes | SessionChange.Focus);
    }

    // Analysis

    public Profile Profile(ImageVariant variant)
    {
        var samples = LineSampler.Sample(variant.Buffer, Line);
        var values = samples.Colors.Select(c => ColorConversion.Value(c, ColorModel, ChannelIndex)).ToArray();
        return new Profile(samples.Positions, values, samples);
    }

    public SpectrumResult Spectrum(Profile profile) =>
        SpectrumAnalyzer.Compute(profile.Values, Math.Max(profile.Samples.Spacing, 1e-9), Window, RemoveMean);

    /// <summary>Frequencies on the chosen axis and magnitudes (in dB when enabled).</summary>
    public (double[] Frequencies, double[] Magnitudes) SpectrumPoints(SpectrumResult s)
    {
        var freqs = FrequencyAxis == FrequencyAxis.CyclesPerPixel ? s.CyclesPerPixel : s.CyclesPerLine;
        var mags = s.Amplitudes.Select(a => LogMagnitude ? Math.Max(20 * Math.Log10(Math.Max(a, 1e-12)), DbFloor) : a).ToArray();
        return (freqs, mags);
    }

    public const double DbFloor = -100.0;

    // Export

    public void ExportFocused(Window? owner)
    {
        var variant = FocusedVariant;
        string baseName = Path.GetFileNameWithoutExtension(FileName);
        string suffix = variant.IsOriginal ? "original" : variant.Name.Replace(" → ", "_");
        foreach (char c in Path.GetInvalidFileNameChars()) suffix = suffix.Replace(c, '-');
        var dialog = new SaveFileDialog
        {
            Filter = "PNG image|*.png",
            DefaultExt = ".png",
            FileName = $"{baseName}-{suffix}.png",
            Title = "Export Image",
        };
        if (dialog.ShowDialog(owner) != true) return;
        try
        {
            var encoder = new PngBitmapEncoder();
            encoder.Frames.Add(BitmapFrame.Create(variant.Bitmap));
            using var stream = File.Create(dialog.FileName);
            encoder.Save(stream);
        }
        catch (Exception ex)
        {
            Error?.Invoke($"Could not write {Path.GetFileName(dialog.FileName)}: {ex.Message}");
        }
    }
}

/// <summary>
/// The session of the most recently activated document window, for the inspector window (which is itself
/// the active window while being used).
/// </summary>
public static class ActiveSession
{
    private static Session? _session;

    public static event Action? Changed;

    public static Session? Current
    {
        get => _session;
        set { if (_session != value) { _session = value; Changed?.Invoke(); } }
    }
}
